pkg/services/hcloud/server/server.go

```diff
@@ -191,6 +248,411 @@ func (s *Service) handleBootStateUnset(ctx context.Context) (reconcile.Result, e
+			"ImageURLCommand", s.scope.ImageURLCommand,
+			"exitStatus", exitStatus,
+			"stdouStderr", stdouStderr)
+		s.scope.SetError(msg, capierrors.CreateMachineError)
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return reconcile.Result{}, nil
+	}
+	conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+		string(hm.Status.BootState), clusterv1.ConditionSeverityInfo,
+		"hcloud-image-url-command started")
+	hm.SetBootState(infrav1.HCloudBootStateWaitForImageURLCommandThenRebootAfterImageURLCommand)
+	return reconcile.Result{RequeueAfter: 55 * time.Second}, nil
+}
+
+func (s *Service) handleBootStateWaitForImageURLCommandThenRebootAfterImageURLCommand(ctx context.Context, server *hcloud.Server) (res reconcile.Result, err error) {
+	hm := s.scope.HCloudMachine
+	updateHCloudMachineStatusFromServer(hm, server)
+
+	hcloudSSHClient, err := s.getSSHClient(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getSSHClient failed (wait for image-url-command): %w", err)
+	}
+
+	state, logFile, err := hcloudSSHClient.StateOfImageURLCommand()
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("StateOfImageURLCommand failed: %w", err)
+	}
+
+	duration := time.Since(hm.Status.BootStateSince.Time)
+	// Please keep the number (7) in sync with the docstring of ImageURL.
+	if duration > 7*time.Minute {
+		// timeout. Something has failed.
+		msg := fmt.Sprintf("ImageURLCommand timed out after %s. Deleting machine",
+			duration.Round(time.Second).String())
+		err = errors.New(msg)
+		s.scope.Logger.Error(err, "",
+			"logFile", logFile)
+		s.scope.SetError(msg, capierrors.CreateMachineError)
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return reconcile.Result{}, nil
+	}
+
+	switch state {
+	case sshclient.ImageURLCommandStateRunning:
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityInfo,
+			"hcloud-image-url-command still running")
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+
+	case sshclient.ImageURLCommandStateFinishedSuccessfully:
+		rebootAction, err := s.scope.HCloudClient.Reboot(ctx, server)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("reboot after ImageURLCommand failed: %w",
+				err)
+		}
+		hm.Status.ActionIDRebootAfterImageURLCommand = rebootAction.ID
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityInfo,
+			"hcloud-image-url-command finished successfully")
+		hm.SetBootState(infrav1.HCloudBootStateWaitForRebootAfterImageURLCommandThenBootToRealOS)
+		return reconcile.Result{RequeueAfter: requeueImmediately}, nil
+
+	case sshclient.ImageURLCommandStateFinishedFailed:
+		msg := "ImageURLCommand failed. Deleting machine"
+		err = errors.New(msg)
```

Comment:
I think you can inline this and also put below lines in one.. Is there a reason to split it up across three lines? Here and in the same way above in another case.

Reply:
done.